driver-core/src/main/com/mongodb/internal/connection/ProtocolHelper.java

```diff
@@ -215,13 +218,17 @@ static void encodeMessage(final RequestMessage message, final BsonOutput bsonOut
         }
     }
 
+    private static final List<Integer> NOT_MASTER_CODES = asList(10107, 13435);
+    private static final List<Integer> RECOVERING_CODES = asList(11600, 11602, 13436, 189, 91);
     private static MongoException createSpecialException(final BsonDocument response, final ServerAddress serverAddress,
                                                          final String errorMessageFieldName) {
-        if (ErrorCategory.fromErrorCode(getErrorCode(response)) == ErrorCategory.EXECUTION_TIMEOUT) {
-            return new MongoExecutionTimeoutException(getErrorCode(response), getErrorMessage(response, errorMessageFieldName));
-        } else if (getErrorMessage(response, errorMessageFieldName).startsWith("not master")) {
+        int errorCode = getErrorCode(response);
+        String errorMessage = getErrorMessage(response, errorMessageFieldName);
+        if (ErrorCategory.fromErrorCode(errorCode) == ErrorCategory.EXECUTION_TIMEOUT) {
+            return new MongoExecutionTimeoutException(errorCode, errorMessage);
+        } else if (errorMessage.startsWith("not master") || NOT_MASTER_CODES.contains(errorCode)) {
             return new MongoNotPrimaryException(serverAddress);
-        } else if (getErrorMessage(response, errorMessageFieldName).startsWith("node is recovering")) {
+        } else if (errorMessage.startsWith("node is recovering") || RECOVERING_CODES.contains(errorCode)) {
```

Comment:
Re-reading the spec, I see this prose:

> The error is considered a "node is recovering" error if the substrings "node is recovering" or "not master or secondary" are anywhere in the error message. Otherwise, if the substring "not master" is in the error message it is a "not master" error.

Let's slip a change to compare the error messages as specified in here as well.

Reply:
Flipped the order and now check if the string contains the phrases.